File: 设计模式/程序/16-InterpreterPattern/InterpreterPattern(C++)/src/Implement.cpp
#include "Implement.h"
#include <iostream>
#include <string>


namespace InterpreterPattern
{
    TerminalExpression::TerminalExpression(const std::string& data) : _data(data) 
    {}

    bool TerminalExpression::Interpreter(const std::string& context)  
    {
        if (context.find(this->_data) != -1)
            return true;
        else return false;
    }



    OrExpression::OrExpression(Expression* expression1, Expression* expression2):
        _expression1(expression1), _expression2(expression2)
    {}

    bool OrExpression::Interpreter(const std::string& context)  
    {
        return this->_expression1->Interpreter(context) || 
               this->_expression2->Interpreter(context);
    }



    AndExpression::AndExpression(Expression* expression1, Expression* expression2):
        _expression1(expression1), _expression2(expression2)
    {}

    bool AndExpression::Interpreter(const std::string& context)  
    {
        return this->_expression1->Interpreter(context) &&
               this->_expression2->Interpreter(context);
    }

    
}

File: 设计模式/程序/16-InterpreterPattern/InterpreterPattern(C#)/Implement.cs
using System;

namespace InterpreterPattern
{
    /* 实现类 */
    public class TerminalExpression : Expression
    {
        public TerminalExpression(string data)
        {
            this.data = data;
        }
        public bool Interpret(string context)
        {
            if (context.Contains(this.data))
                return true;
            return false;
        }

        private string data;
    }

    public class OrExpression : Expression
    {
        public OrExpression(Expression expression1, Expression expression2)
        {
            this.expression1 = expression1;
            this.expression2 = expression2;
        }
        public bool Interpret(string context)
        {
            return this.expression1.Interpret(context) || this.expression2.Interpret(context);
        }

        private Expression expression1;
        private Expression expression2;
    }

    public class AndExpression : Expression
    {
        public AndExpression(Expression expression1, Expression expression2)
        {
            this.expression1 = expression1;
            this.expression2 = expression2;
        }
        public bool Interpret(string context)
        {
            return this.expression1.Interpret(context) && this.expression2.Interpret(context);
        }

        private Expression expression1;
        private Expression expression2;
    }
}